File: RayTracerVBSingleProj/Thing.vb
Public MustInherit Class Thing
    Public ReadOnly Property Surface As SurfaceTexture

    Public Sub New(ByVal surface As SurfaceTexture)
        Me.Surface = surface
    End Sub

    MustOverride Function CalculateIntersection(ByVal withRay As Ray) As Intersection
    MustOverride Function CalculateNormal(ByVal toSurfacePosition As Vector3) As Vector3
End Class

File: RayTracerCSSingleProj/Thing.cs
namespace RayTracer
{
    public abstract class Thing
    {
        public SurfaceTexture Surface { get; private set; }
        public abstract Intersection CalculateIntersection(Ray withRay);
        public abstract Vector3 CalculateNormal(Vector3 toSurfacePosition);

        public Thing(SurfaceTexture surface)
        {
            Surface = surface;
        }
    }
}
